cmd/registry/cmd/get/get.go

```diff
@@ -15,90 +15,329 @@
-				err2 = core.GetAPI(ctx, client, api, core.PrintAPIDetail)
-			} else if deployment, err := names.ParseDeployment(args[0]); err == nil {
-				err2 = core.GetDeployment(ctx, client, deployment, core.PrintDeploymentDetail)
-			} else if deployment, err := names.ParseDeploymentRevision(args[0]); err == nil {
-				err2 = core.GetDeploymentRevision(ctx, client, deployment, core.PrintDeploymentDetail)
-			} else if version, err := names.ParseVersion(args[0]); err == nil {
-				err2 = core.GetVersion(ctx, client, version, core.PrintVersionDetail)
-			} else if spec, err := names.ParseSpec(args[0]); err == nil {
-				// for specs, these options are synonymous
-				if getPrintedContents || getRawContents {
-					err2 = core.GetSpec(ctx, client, spec, true, core.WriteSpecContents)
-				} else {
-					err2 = core.GetSpec(ctx, client, spec, false, core.PrintSpecDetail)
-				}
-			} else if spec, err := names.ParseSpecRevision(args[0]); err == nil {
-				// for specs, these options are synonymous
-				if getPrintedContents || getRawContents {
-					err2 = core.GetSpecRevision(ctx, client, spec, true, core.WriteSpecContents)
-				} else {
-					err2 = core.GetSpecRevision(ctx, client, spec, false, core.PrintSpecDetail)
-				}
-			} else if artifact, err := names.ParseArtifact(args[0]); err == nil {
-				if getPrintedContents {
-					err2 = core.GetArtifact(ctx, client, artifact, true, core.PrintArtifactContents)
-				} else if getRawContents {
-					err2 = core.GetArtifact(ctx, client, artifact, true, core.WriteArtifactContents)
-				} else {
-					err2 = core.GetArtifact(ctx, client, artifact, false, core.PrintArtifactDetail)
-				}
-			} else {
-				log.Errorf(ctx, "Unsupported entity %+v", args)
-			}
-			if err2 != nil {
-				log.FromContext(ctx).WithError(err2).Errorf("Failed to get resource")
+				return err
 			}
+			h := &getHandler{
+				ctx:         ctx,
+				client:      client,
+				adminClient: adminClient,
+				writer:      cmd.OutOrStdout(),
+				name:        args[0],
+				filter:      filter,
+				output:      output,
+			}
+			err = h.traverse()
+			if err != nil {
+				return err
+			}
+			return h.write()
 		},
 	}
 
-	cmd.Flags().BoolVar(&getContents, "contents", false, "Get resource contents if available (deprecated)")
-	cmd.Flags().BoolVar(&getRawContents, "raw", false, "Get raw resource contents if available")
-	cmd.Flags().BoolVar(&getPrintedContents, "print", false, "Print resource contents if available")
+	cmd.Flags().StringVar(&filter, "filter", "", "Filter selected resources")
+	cmd.Flags().StringVarP(&output, "output", "o", "name", "Output type (name, yaml, contents)")
 	return cmd
 }
+
+type getHandler struct {
+	ctx         context.Context
+	client      connection.RegistryClient
+	adminClient connection.AdminClient
+	writer      io.Writer
+	name        string
+	filter      string
+	output      string
+	results     []interface{} // result values to be returned in a single message
+}
```

Comment:
Which of these actually need to be fields on a struct, as opposed to parameters of a function? In a 60 second time box I was able to figure out that `traverse()` and `write()` are the two methods meant for direct use, and the others are helpers. It also seems like everything except `writer` and `results` are just used as inputs. Why isn't this just two functions?

The context for example, definitely shouldn't be a field on a struct. Each function that needs a context should accept it as an argument.

https://pkg.go.dev/context#pkg-overview
https://google.github.io/styleguide/go/decisions#contexts

Reply:
This is struct collects state that is captured in closures passed to the `core.List*` and `core.Get*` functions.

Reply:
#913